runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -659,83 +754,196 @@ private void setCurrentOutputWatermark(long currentOutputWatermark) {
+          StateTags.bag("bundle-buffer-tag", new TaggedKvCoder(tagToLabel, coderMapping));
+      bufferState = stateInternals.state(StateNamespaces.global(), bufferTag);
+    }
+
+    /**
+     * Transform TupleTags to labels. Integer is easier to serialize than TupleTag.
+     * Returns a BiMap for easier reversal query.
+     */
+    private static BiMap<TupleTag<?>, Integer> transformTupleTagsToLabels(
+        TupleTag<?> mainTag,
+        Map<TupleTag<?>, OutputTag<WindowedValue<?>>> mapping) {
+
+      BiMap<TupleTag<?>, Integer> tagToLabel = HashBiMap.create();
+      int count = 0;
+      tagToLabel.put(mainTag, count++);
+      for (TupleTag<?> key : mapping.keySet()) {
+        if (!tagToLabel.containsKey(key)) {
+          tagToLabel.put(key, count++);
         }
-      };
+      }
+      return tagToLabel;
+    }
+
+    void openBuffer() {
+      this.openBuffer = true;
+    }
+
+    void closeBuffer() {
+      this.openBuffer = false;
+    }
+
+    @Override
+    public <T> void output(TupleTag<T> tag, WindowedValue<T> value) {
+      if (!openBuffer) {
+        emit(tag, value);
+      } else {
+        bufferState.add(KV.<Integer, WindowedValue<?>>of(tagToLabel.get(tag), value));
+      }
+    }
+
+    /**
+     * Flush elements of bufferState to Flink Output. This method can't be invoke in
+     * {@link #snapshotState(StateSnapshotContext)}
+     */
+    void flushBuffer() {
+      for (KV<Integer, WindowedValue<?>> tuple2 : bufferState.read()) {
+        emit(tagToLabel.inverse().get(tuple2.getKey()), (WindowedValue) tuple2.getValue());
+      }
+      bufferState.clear();
+    }
+
+    private <T> void emit(TupleTag<T> tag, WindowedValue<T> value) {
+      if (tag.equals(mainTag)) {
+        // with tagged outputs we can't get around this because we don't
+        // know our own output type...
+        @SuppressWarnings("unchecked")
+        WindowedValue<OutputT> castValue = (WindowedValue<OutputT>) value;
+        output.collect(new StreamRecord<>(castValue));
+      } else {
+        @SuppressWarnings("unchecked")
+        OutputTag<WindowedValue<T>> outputTag = (OutputTag) mapping.get(tag);
+        output.collect(outputTag, new StreamRecord<>(value));
+      }
     }
   }
 
   /**
+   * Coder for KV of label and value. It will be serialized in Flink checkpoint.
+   */
+  private static class TaggedKvCoder extends StructuredCoder<KV<Integer, WindowedValue<?>>> {
+
+    private BiMap<TupleTag<?>, Integer> tagToLabel;
+    private Map<TupleTag<?>, Coder<WindowedValue<?>>> coderMapping;
+
+    TaggedKvCoder(
+        BiMap<TupleTag<?>, Integer> tagToLabel,
+        Map<TupleTag<?>, Coder<WindowedValue<?>>> coderMapping) {
+      this.tagToLabel = tagToLabel;
+      this.coderMapping = coderMapping;
+    }
+
+    @Override
+    public void encode(KV<Integer, WindowedValue<?>> kv, OutputStream out)
+        throws IOException {
+      Coder<WindowedValue<?>> coder = coderMapping.get(tagToLabel.inverse().get(kv.getKey()));
```

Comment:
If we always only need the `inverse()` we could get the inverse in the constructor.